### ISSUE TITLE
OOM Error on A100 GPU While Fine-Tuning GenCast with 0.25° Resolution Data

### ISSUE THREAD

**Issue by Wangsherpa** [google-deepmind/graphcast#171]
**Description**
I'm encountering an out-of-memory (OOM) error when attempting to run gradient computation using the **`gencast_mini_demo`** notebook on a single A100 (80GB) GPU. The issue occurs after switching to the 0.25° HRES dataset and the corresponding model checkpoint.

---

**Setup**:

* **Notebook**: `gencast_mini_demo.ipynb`
* **Model Checkpoint**: `GenCast 0p25deg Operational <2022`
* **Dataset**: `hres_date-2022-01-01_res-0.25_levels-13_steps-01.nc`
* **Resolution**: 0.25°, 13 levels
* **Hardware**: NVIDIA A100 (80GB), 8 GPU (aws p4de.24xlarge cluster)
* **Environment**: Python 3.11.11, JAX 0.6.2

---

**Code causing OOM**:

```python
loss, diagnostics, next_state, grads = grads_fn_jitted(
    params=params,
    state=state,
    inputs=train_inputs,
    targets=train_targets,
    forcings=train_forcings)

mean_grad = np.mean(jax.tree_util.tree_flatten(jax.tree_util.tree_map(lambda x: np.abs(x).mean(), grads))[0])
print(f"Loss: {loss:.4f}, Mean |grad|: {mean_grad:.6f}")
```

---

The error trace points to:
  ```
Execution of replica 0 failed: RESOURCE_EXHAUSTED: Out of memory while trying to allocate 765258926192 bytes. [tf-allocator-allocation-error='']
  ```

**Questions**:
1. Is the 0.25° operational model intended to be fine-tuned on GPU setups like A100?
2. Is the training loop provided in the demo notebook compatible with multi-GPU setups using jax.pmap or other strategies? If not, are there guidelines for enabling it?

**Comment by andrewlkd** [comment]
> Would migrating to a TPU setup be the recommended approach?

A TPU setup will certainly help reduce memory requirements because you'll be able to make use of the splash attention kernel. Training on a GPU should still be possible but you may have to shard or employ some other memory saving techniques.

> What would be the minimal TPU configuration required to successfully run full training/fine-tuning at this resolution?

Unsharded GenCast can be trained on a single TPUv5p. 

> Is there any suggested way to fine-tune gencast 0.25 model on GPU, even if it's constrained or experimental?

Other than sharding the model, you could consider things like freezing parts of the network and only training e.g. the decoding heads. I can't say for certain whether this would reduce memory sufficiently and how much you'd have to freeze but I'd probably start there.

Hope this helps,

Andrew

**Comment by Wangsherpa** [comment]
Thanks so much Andrew for the detailed response — this really helps clarify the options. I’ll look into sharding and possibly freezing parts of the model as a starting point for GPU-based fine-tuning.
